core/src/mindustry/ai/types/FormationAI.java

```diff
@@ -46,10 +48,13 @@ public void updateUnit(){
             unit.lookAt(unit.vel.angle());
         }
 
-        Vec2 realtarget = vec.set(target).add(leader.vel);
+        Vec2 next = vec.set(target).add(leader.vel);
+        if(unit.isGrounded() && !unit.within(leader, 4 * formation.pattern.spacing)){
+            next = vec.set(Astar.pathfind(unit.tileOn(), leader.tileOn(), FormationAI::pathfindCost, t -> unit.canPass(t.x, t.y)).first());
```

Comment:
You cannot simply use A* to pathfind here. This is **horribly** inefficient and spews garbage every frame.

Reply:
but nothing else would work unless... tell the unit to move next to walls until it's free...

Reply:
won't work really. But can we at least calculate the path once and follow it unless it ends or we are close enough

Reply:
> calculate the path once

- What if the player unit is inaccessible?
- What if it's on the other side of the map? 
- What if it moves behind an obstacle?
- What if the formation unit gets shoved or otherwise displaced somewhere?
- What if this is running on any number of devices that can't pathfind from point A to point B in <16ms?
- What if there are many units that need to calculate the path?